### PR TITLE
fix the issue of #772 (Can't acquire memory from spark Memory Manager)

### PR DESCRIPTION
**Pull request by JkSelf**
## What changes were proposed in this pull request?

The init method should be synchronized and the OapRuntime object is created only when the rt is null to make sure the OapRuntime object be created only one time.

## How was this patch tested?

pass the unit test of OapRuntimeSuite

### CODE REVIEW

**Comment by JkSelf** [comment]
@LuciferYang @carsonwang @gczsjdy please help review again, thanks.

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapRuntimeSuite.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    sql(s"""CREATE TEMPORARY VIEW oap_sort_opt_table (a INT, b STRING)
+            | USING oap
+            | OPTIONS (path '$path1')""".stripMargin)
+  }
+
+  override def afterEach(): Unit = {
+    super.afterEach()
+  }
+
+  test("OapRuntime Test") {
```
Comment (LuciferYang) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapRuntimeSuite.scala]: Add issue number and descriptions to test name ?

Comment (LuciferYang) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapRuntimeSuite.scala]: We add `synchronized` to solve the OapRuntime concurrent initialization ?
The scene is MemoryManager require memory from mem pool many times not one ? 
Can we assert the expectation of the remaining memory of MemoryPool after the OapRuntime initialization?

Comment (gczsjdy) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapRuntimeSuite.scala]: In `LocalCluster` mode, the current non-synchronized OapRuntime `init` will create the memory problem. This only occurs when multiple executors run on 1 host.

Comment (gczsjdy) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapRuntimeSuite.scala]: @JkSelf We can refine the test suite, assert something like @LuciferYang suggested, it's more clear.

Comment (JkSelf) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapRuntimeSuite.scala]: @LuciferYang @gczsjdy  thanks for your review. 
@LuciferYang yes, currently, the OapRuntime will be created many times when occurring concurrent situation, which may cause the "Can't acquire memory from spark Memory Manager" issue in MemoryManager. So we add the synchronized to enable it be created only once. For the assert, we check whether the OapRuntime  is the same object when two thread call the init method simultaneously. Please help review, thanks.

File: src/test/scala/org/apache/spark/sql/oap/rpc/OapRuntimeSuite.scala [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            oapruntime(i) = OapRuntime.getOrCreate
+          }
+        })
+      }
+    } finally {
+      threadPool.shutdown()
+    }
+    assert(oapruntime(0) == oapruntime(1))
+  }
+}
```
Comment (LuciferYang) [inline src/test/scala/org/apache/spark/sql/oap/rpc/OapRuntimeSuite.scala]: should add a empty line at the end of file.

File: src/test/scala/org/apache/spark/sql/oap/rpc/OapRuntimeSuite.scala [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    val oapruntime = new Array[OapRuntime](2)
+    val threadPool: ExecutorService = Executors.newFixedThreadPool(2)
+    try {
+      for (i <- 0 to 1) {
+        threadPool.execute(new Runnable {
+          override def run(): Unit = {
+            oapruntime(i) = OapRuntime.getOrCreate
+          }
+        })
+      }
```
Comment (gczsjdy) [inline src/test/scala/org/apache/spark/sql/oap/rpc/OapRuntimeSuite.scala]: We should wait for the 2 threads to complete in `try`

File: src/test/scala/org/apache/spark/sql/oap/rpc/OapRuntimeSuite.scala [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.oap
+
+import java.util.concurrent.{ Executors, ExecutorService}
+
+import org.apache.spark.sql.QueryTest
+import org.apache.spark.sql.test.oap.SharedOapLocalClusterContext
+
```
Comment (gczsjdy) [inline src/test/scala/org/apache/spark/sql/oap/rpc/OapRuntimeSuite.scala]: :nit eliminate a line

**Comment by carsonwang** [comment]
Thanks @JkSelf . Merging..

**Comment by LuciferYang** [comment]
@carsonwang  does the v0.4 need this patch also ?

**Comment by gczsjdy** [comment]
@LuciferYang Yes, I have pushed to branch-0.4

**Comment by LuciferYang** [comment]
@gczsjdy thx~

**Pull request merged**